src/esp/sim/ReplayBatchRenderer.h

```diff
@@ -54,42 +147,103 @@ class ReplayBatchRenderer {
-  // ReplayBatchRendererConfiguration::sensorSpecifications, for the specified
-  // environment's keyframe. See also setEnvironmentKeyframe.
-  void setSensorTransformsFromKeyframe(int envIndex, const std::string& prefix);
+  Magnum::Vector2i doSensorSize(unsigned envIndex) override;
 
- private:
-  scene::SceneNode* loadAndCreateRenderAssetInstance(
-      int envIndex,
+  esp::gfx::replay::Player& doPlayerFor(unsigned envIndex) override;
+
+  void doSetSensorTransform(unsigned envIndex,
+                            const std::string& sensorName,
+                            const Mn::Matrix4& transform) override;
+
+  void doSetSensorTransformsFromKeyframe(unsigned envIndex,
+                                         const std::string& prefix) override;
+
+  void doRender(Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+                    imageViews) override;
+
+  void doRender(Magnum::GL::AbstractFramebuffer& framebuffer) override;
+
+  gfx::replay::NodeHandle loadAndCreateRenderAssetInstance(
+      unsigned envIndex,
       const assets::AssetInfo& assetInfo,
       const assets::RenderAssetInstanceCreationInfo& creation);
 
   std::vector<EnvironmentRecord> envs_;
 
-  std::unique_ptr<assets::ResourceManager> resourceManager_ = nullptr;
+  std::unique_ptr<assets::ResourceManager> resourceManager_;
 
   scene::SceneManager::uptr sceneManager_ = nullptr;
 
   gfx::WindowlessContext::uptr context_ = nullptr;
   std::shared_ptr<gfx::Renderer> renderer_ = nullptr;
 
-  ReplayBatchRendererConfiguration config_;
+  ReplayRendererConfiguration config_;
+
+  ESP_SMART_POINTERS(ReplayRenderer)
+};
+
+class ReplayBatchRenderer : public AbstractReplayRenderer {
+ public:
+  explicit ReplayBatchRenderer(const ReplayRendererConfiguration& cfg);
+
+  ~ReplayBatchRenderer() override;
+
+ private:
+  void doPreloadFile(Corrade::Containers::StringView filename) override;
+
+  unsigned doEnvironmentCount() const override;
+
+  Magnum::Vector2i doSensorSize(unsigned envIndex) override;
+
+  esp::gfx::replay::Player& doPlayerFor(unsigned envIndex) override;
+
+  void doSetSensorTransform(unsigned envIndex,
+                            const std::string& sensorName,
+                            const Mn::Matrix4& transform) override;
+
+  void doSetSensorTransformsFromKeyframe(unsigned envIndex,
+                                         const std::string& prefix) override;
+
+  void doRender(Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+                    imageViews) override;
+
+  void doRender(Magnum::GL::AbstractFramebuffer& framebuffer) override;
+
+  /* If standalone_ is true, renderer_ contains a RendererStandalone. Has to be
+     before the EnvironmentRecord array because Player calls
+     gfx_batch::Renderer::clear() on destruction. */
+  bool standalone_;
+  Corrade::Containers::Pointer<esp::gfx_batch::Renderer> renderer_;
+
+  // TODO pimpl all this?
+  struct EnvironmentRecord {
```

Comment:
This use of a virtual interface per env doesn't strike me as very data-oriented. I was hoping to only see virtual-call overhead at the renderer level, not per environment (remember we may eventually have thousands of envs in the batch use case).

Reply:
Hmm. Are you more concerned about the size of the `EnvironmentRecord` struct (due to the vtables and other overhead), or about the one-virtual-call-per-transformation-update overhead?

The former could be solved relatively easily by having the Player remember `envId`, because that's the only single thing that differs between the instances, and I did it this way because it was the most direct change from the `std::function`s that were there originally (where *each* `std::function` was a virtual interface on its own). This change isn't hard, won't affect the public API in any way, and I propose doing it in the next iteration, as this PR is big enough already :)

The latter, one-virtual-call-per-transformation-update would have to be solved in the Player itself, where it would have to put the (parsed) transforms into some memory location and then submit them all at once with a single (virtual) call. The batch renderer is already prepared for such workflow, as it [exposes the transformation array directly](https://aihabitat.org/docs/habitat-sim/classesp_1_1gfx__batch_1_1Renderer.html#acfb2439f0fe5e5e7397eaf3d517a49ab), for the classic renderer it'd be about looping over the passed array and updating. That's a more substantial change, but still doable, and I think it should go into a standalone PR on its own.

Reply:
>the one-virtual-call-per-transformation-update overhead?
This is my concern (not the struct size).

>That's a more substantial change, but still doable, and I think it should go into a standalone PR on its own.
Yep, agree it's a substantial change. We should do this in the future.

>it would have to put the (parsed) transforms into some memory location and then submit them all at once with a single (virtual) call.
Consider that the Python caller will have gathered all keyframes before passing them on to the batch renderer. So we should change the Python API to pass all keyframes at once (thus no need to use additional memory).

Reply:
Just to clarify, I meant having something like

```cpp
virtual void setNodeTransforms(
    StridedArrayView1D<const NodeHandle> nodes,
    StridedArrayView1D<const Magnum::Vector3> translations,
    StridedArrayView1D<const Magnum::Quaternion> rotations);
```

in the `AbstractPlayerImplementation`, thus having exactly one virtual call per frame. Or whichever other variant where the data get passed in as a reference to some regularly shaped memory.

Reply:
Ultimately, when we go to optimize the batch replay renderer, I suspect we'll just avoid using the Player base class. A keyframe is a straightforward data structure to map to a sequence of new instance and light transforms (and the other ops like instance creation); we should just write that code directly. I don't care that much about sharing this Keyframe-parsing logic between the two backends. I'm okay with some code duplication. Sorry I didn't bring this up before; I do think our solution here in this PR (a shared Player base class) is fine for the near term.

Reply:
Okay, so -- anything left to do here regarding this point? :) As I see it, it's all for subsequent PRs.

Reply:
Nope, all for subsequent PRs.